### PR TITLE
Performance tests for loading register data

### LINKED ISSUE #2066: [App Performance] Profile the time it takes for the app to load the configs and patient data from the local DB.
**Describe the enhancement**
- The FHIR core performs a comprehensive initial sync. This sync gets the following items. 
    - Application Composition resource (Based on the Application ID)
    - Application Congis referenced on the Composition
    - Patient data based on the sync configs

**Implementation**
- [ ] Review the time and device resources it takes for the application to query the local DB to fetch the following pieces of data. 
    - [ ] The time and device resources to get the Configs used by the app views. 
    - [ ] The time and device resources to get the Patient data. Here we can look at 3 different resources
        - [ ] Patient resource.
        - [ ] Group resource.
        - [ ] Task resource. 
        - [ ] Observation resource.
        - [ ] CarePlan resource.
        - [ ] MeasureReports resource.

 **Device configuration**
- RAM 3GB
- ROM 64GB 

Expected results. 
- Profiling documentation can be done by recording the following
    - [ ] Time and device resources it takes for each data fetch. 
    - [ ] Amount of resources used by the device for each request.  i.e RAM & CPU

### PR DESCRIPTION
**Pull request by ekigamba**
**IMPORTANT: Where possible all PRs must be linked to a Github issue**

Fixes #2066

**Engineer Checklist**
- [ ] I have written **Unit tests** for any new feature(s) and edge cases for bug fixes
- [ ] I have added any strings visible on UI components to the `strings.xml` file
- [ ] I have updated the  [CHANGELOG.md](./CHANGELOG.md) file for any notable changes to the codebase
- [ ] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the project's style guide
- [ ] I have built and run the FHIRCore app to verify my change fixes the issue and/or does not break the app 
- [ ] I have checked that this PR does NOT introduce **breaking changes** that require an update to **_Content_** and/or **_Configs_**? _If it does add a sample here or a link to exactly what changes need to be made to the content._

**Code Reviewer Checklist**
- [x] I have verified **Unit tests** have been written for any new feature(s) and edge cases
- [x] I have verified any strings visible on UI components are in the `strings.xml` file
- [ ] I have verifed the [CHANGELOG.md](./CHANGELOG.md) file has any notable changes to the codebase
- [ ] I have verified the solution has been implemented in a configurable and generic way for reuseable components
- [ ] I have built and run the FHIRCore app to verify the change fixes the issue and/or does not break the app

### CODE REVIEW

File: android/quest/src/androidTest/java/org/smartregister/fhircore/performance/HouseholdRegisterPerformanceTests.kt [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    // Copy over the db
+    Files.copy(
+      resourcesDbInputStream,
+      Path("/data/data/${appContext.packageName}/databases/resources.db"),
+    )
+  }
+
+  @Test
+  fun benchmarkPage0() {
+    benchmarkingFunctionality(0)
```
Comment (pld) [inline android/quest/src/androidTest/java/org/smartregister/fhircore/performance/HouseholdRegisterPerformanceTests.kt:86]: can this helper function return the time it took? then in here we can put it in an assert, like
```suggestion
    val secondsToRun = benchmarkingFunctionality(0)
    assertTrue(secondsToRun < 180)
```
or if these all take around the same time to run, we can put the assert in the helper function

Comment (ekigamba) [reply android/quest/src/androidTest/java/org/smartregister/fhircore/performance/HouseholdRegisterPerformanceTests.kt:46]: The benchmarks don't return the timing results. The timings results are saved in a file. These are provided as median, max, min and all the individual iteration runtimes. 

For us to assert, we need to run a gradle task to parse this file and compare the times for each test result

File: android/quest/build.gradle.kts [hunk 2]
```diff
@@ -300,6 +305,10 @@ tasks.withType<Test> {
   minHeapSize = "4608m"
   maxHeapSize = "4608m"
   maxParallelForks = (Runtime.getRuntime().availableProcessors() / 2).takeIf { it > 0 } ?: 1
+
+  if (!name.toLowerCase().contains("performance")) {
```
Comment (ekigamba) [inline android/quest/build.gradle.kts:309]: Test this out to see if it works

File: android/quest/build.gradle.kts [hunk 3]
```diff
@@ -361,8 +370,79 @@ dependencies {
   androidTestImplementation(libs.ui.test.junit4)
   androidTestImplementation(libs.hilt.android.testing)
   androidTestImplementation(libs.mockk.android)
+  androidTestImplementation("androidx.benchmark:benchmark-junit4:1.1.1")
```
Comment (ekigamba) [inline android/quest/build.gradle.kts:373]: Move this to `libs.versions.toml`

File: android/quest/build.gradle.kts [hunk 4]
```diff
@@ -361,8 +370,79 @@ dependencies {
...
   ktlint(libs.ktlint.main) {
     attributes { attribute(Bundling.BUNDLING_ATTRIBUTE, objects.named(Bundling.EXTERNAL)) }
   }
   ktlint(project(":linting"))
 }
+
+task("evaluatePerformanceBenchmarkResults") {
```
Comment (ekigamba) [inline android/quest/build.gradle.kts:380]: Add documentation for this

File: android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/register/RegisterRepository.kt [hunk 5]
```diff
@@ -107,7 +107,11 @@ constructor(
         .runCatching {
           withContext(dispatcherProvider.io()) { fhirEngine.search<Resource>(search) }
         }
-        .onFailure { Timber.e(it, "Error retrieving resources. Empty list returned by default") }
+        .onFailure {
+          Timber.e(it, "Error retrieving resources. Empty list returned by default")
+          System.out.println("Error occurred")
+          it.printStackTrace()
```
Comment (ekigamba) [inline android/engine/src/main/java/org/smartregister/fhircore/engine/data/local/register/RegisterRepository.kt:113]: Remove these two

File: android/quest/src/androidTest/java/org/smartregister/fhircore/performance/HouseholdRegisterPerformanceTests.kt (hunk 1 above)
Comment (pld) [reply android/quest/src/androidTest/java/org/smartregister/fhircore/performance/HouseholdRegisterPerformanceTests.kt:46]: cool

**Review by pld — DISMISSED** [review]

File: .github/workflows/performance-tests.yml [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+ FHIRCORE_USERNAME: ${{ secrets.FHIRCORE_USERNAME }}
+ FHIRCORE_ACCESS_TOKEN: ${{ secrets.FHIRCORE_ACCESS_TOKEN }}
+ CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
+
+jobs:
+  quest-performance-tests:
+    runs-on: macos-latest
+    strategy:
+      matrix:
+        api-level: [28]
```
[the remark this answers is not in the document]
Comment (Kigamba) [reply .github/workflows/performance-tests.yml:19]: Yea

**Review by pld — DISMISSED** [review]

File: android/quest/build.gradle.kts (hunk 3 above)
Comment (pld) [reply android/quest/build.gradle.kts]: does this still need to be moved?

File: android/quest/build.gradle.kts (hunk 4 above)
Comment (pld) [reply android/quest/build.gradle.kts:383]: inline docs are done, I'm going to add docs on perf tests overall...

**Review by pld — APPROVED** [review]

File: docs/engineering/android-app/introduction/readme.mdx [hunk 7]
```diff
@@ -52,8 +52,24 @@ The application uses FHIR Engine APIs from Google's [Android FHIR SDK](https://g
...
 The application syncs particular resources (conventionally, Composition and Binary) from the HAPI FHIR server to configure the app. The configurations control application workflows as well as the look and feel of the app.
 
-### Tests
+### Testing
 
-This FHIR Core repository includes:
-1. Unit tests
-2. UI and integration tests
+The OpenSRP Android app includes automated style, coverage, unit, user inteface / integartion, and performance testing. All tests are run through github actions on pull request and must be passed for a pull request to be merged without an admin override.
+
```
Comment (Rkareko) [inline docs/engineering/android-app/introduction/readme.mdx:58]: change `must be passed` to `must pass`

File: android/quest/build.gradle.kts (hunk 3 above)
Comment (ekigamba) [reply android/quest/build.gradle.kts]: Yes

File: .github/workflows/performance-tests.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (ekigamba) [reply .github/workflows/performance-tests.yml:19]: Yes

File: android/quest/build.gradle.kts (hunk 4 above)
Comment (ekigamba) [reply android/quest/build.gradle.kts:383]: Thanks for this

File: android/quest/src/androidTest/java/org/smartregister/fhircore/quest/config/ConfigServiceModule.kt [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+package org.smartregister.fhircore.quest.di.config
+
+import dagger.Binds
+import dagger.Module
+import dagger.hilt.InstallIn
+import dagger.hilt.components.SingletonComponent
+import org.smartregister.fhircore.engine.configuration.app.ConfigService
+import org.smartregister.fhircore.quest.QuestConfigService
+
+@InstallIn(SingletonComponent::class)
```
[the remark this answers is not in the document]
Comment (ekigamba) [reply android/quest/src/androidTest/java/org/smartregister/fhircore/quest/config/ConfigServiceModule.kt:26]: Nope, I'm going to work on this

**Pull request merged**